In [83]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt

In [84]:
df=pd.read_csv("/Users/akula/Downloads/sales_data_sample.csv", encoding='unicode_escape')
df

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
4           10159               49     100.00               14  5205.27   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

            ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0      2/24/2003 0:00   Shipped       1         2     2003  ...   
1       5/7/2003 0:00   Shipped       2         5     2003  ...   
2       7/1/2003 0:00   Shipped       3         7     2003  ...   
3      8/25/2003 0:00   Shipped       3         8     2003  ...   
4     10/10/2003 0:00   Shipped       4        10     2003  ...   
...               ...       ...     ...       ...      ...  ...   
2818   12/2/2004 0:00   Shipped       4        12     2004  ...   
2819   1/31/2005 0:00   Shipped       1         1     2005  ...   
2820    3/1/2005 0:00  Resolved       1         3     2005  ...   
2821   3/28/2005 0:00   Shipped       1         3     2005  ...   
2822    5/6/2005 0:00   On Hold       2         5     2005  ...   

                       ADDRESSLINE1  ADDRESSLINE2           CITY STATE  \
0           897 Long Airport Avenue           NaN            NYC    NY   
1                59 rue de l'Abbaye           NaN          Reims   NaN   
2     27 rue du Colonel Pierre Avia           NaN          Paris   NaN   
3                78934 Hillside Dr.           NaN       Pasadena    CA   
4                   7734 Strong St.           NaN  San Francisco    CA   
...                             ...           ...            ...   ...   
2818             C/ Moralzarzal, 86           NaN         Madrid   NaN   
2819                    Torikatu 38           NaN           Oulu   NaN   
2820             C/ Moralzarzal, 86           NaN         Madrid   NaN   
2821          1 rue Alsace-Lorraine           NaN       Toulouse   NaN   
2822             8616 Spinnaker Dr.           NaN         Boston    MA   

     POSTALCODE  COUNTRY TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0         10022      USA       NaN              Yu             Kwai    Small  
1         51100   France      EMEA         Henriot             Paul    Small  
2         75508   France      EMEA        Da Cunha           Daniel   Medium  
3         90003      USA       NaN           Young            Julie   Medium  
4           NaN      USA       NaN           Brown            Julie   Medium  
...         ...      ...       ...             ...              ...      ...  
2818      28034    Spain      EMEA          Freyre            Diego    Small  
2819      90110  Finland      EMEA       Koskitalo           Pirkko   Medium  
2820      28034    Spain      EMEA          Freyre            Diego   Medium  
2821      31000   France      EMEA          Roulet          Annette    Small  
2822      51003      USA       NaN         Yoshido             Juri   Medium  

[2823 rows x 25 columns]

In [85]:
df.shape

(2823, 25)

**There are 2823 rows and 25 columns in the dataset**

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ORDERNUMBER       2823 non-null   int64  
 1   QUANTITYORDERED   2823 non-null   int64  
 2   PRICEEACH         2823 non-null   float64
 3   ORDERLINENUMBER   2823 non-null   int64  
 4   SALES             2823 non-null   float64
 5   ORDERDATE         2823 non-null   object 
 6   STATUS            2823 non-null   object 
 7   QTR_ID            2823 non-null   int64  
 8   MONTH_ID          2823 non-null   int64  
 9   YEAR_ID           2823 non-null   int64  
 10  PRODUCTLINE       2823 non-null   object 
 11  MSRP              2823 non-null   int64  
 12  PRODUCTCODE       2823 non-null   object 
 13  CUSTOMERNAME      2823 non-null   object 
 14  PHONE             2823 non-null   object 
 15  ADDRESSLINE1      2823 non-null   object 
 16  ADDRESSLINE2      302 non-null    object 


In [87]:
df['ORDERDATE']=pd.to_datetime(df['ORDERDATE'])

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2823 entries, 0 to 2822
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   ORDERNUMBER       2823 non-null   int64         
 1   QUANTITYORDERED   2823 non-null   int64         
 2   PRICEEACH         2823 non-null   float64       
 3   ORDERLINENUMBER   2823 non-null   int64         
 4   SALES             2823 non-null   float64       
 5   ORDERDATE         2823 non-null   datetime64[ns]
 6   STATUS            2823 non-null   object        
 7   QTR_ID            2823 non-null   int64         
 8   MONTH_ID          2823 non-null   int64         
 9   YEAR_ID           2823 non-null   int64         
 10  PRODUCTLINE       2823 non-null   object        
 11  MSRP              2823 non-null   int64         
 12  PRODUCTCODE       2823 non-null   object        
 13  CUSTOMERNAME      2823 non-null   object        
 14  PHONE             2823 n

**From above data we have datetime 1 datetype, 2 float datatypes, 7 integer datatypes, 15 object datatypes. We have converted one datatype that is from object data type to date time datatype**

In [89]:
df.describe()

ORDERNUMBER  QUANTITYORDERED    PRICEEACH  ORDERLINENUMBER  \
count   2823.000000      2823.000000  2823.000000      2823.000000   
mean   10258.725115        35.092809    83.658544         6.466171   
min    10100.000000         6.000000    26.880000         1.000000   
25%    10180.000000        27.000000    68.860000         3.000000   
50%    10262.000000        35.000000    95.700000         6.000000   
75%    10333.500000        43.000000   100.000000         9.000000   
max    10425.000000        97.000000   100.000000        18.000000   
std       92.085478         9.741443    20.174277         4.225841   

              SALES                      ORDERDATE       QTR_ID     MONTH_ID  \
count   2823.000000                           2823  2823.000000  2823.000000   
mean    3553.889072  2004-05-11 00:16:49.989373056     2.717676     7.092455   
min      482.130000            2003-01-06 00:00:00     1.000000     1.000000   
25%     2203.430000            2003-11-06 12:00:00     2.000000     4.000000   
50%     3184.800000            2004-06-15 00:00:00     3.000000     8.000000   
75%     4508.000000            2004-11-17 12:00:00     4.000000    11.000000   
max    14082.800000            2005-05-31 00:00:00     4.000000    12.000000   
std     1841.865106                            NaN     1.203878     3.656633   

          YEAR_ID         MSRP  
count  2823.00000  2823.000000  
mean   2003.81509   100.715551  
min    2003.00000    33.000000  
25%    2003.00000    68.000000  
50%    2004.00000    99.000000  
75%    2004.00000   124.000000  
max    2005.00000   214.000000  
std       0.69967    40.187912

**By using describe function we got count, mean, min, percentile, maximum, standard deviation values which are use in further data cleaning and data visualization**

# **HANDLING MISSING VALUES**

In [90]:
df.isnull().sum().sum()

np.int64(5157)

**There are total 5157 missing values we need to replace or remove the values to make data more accurate for future data visualization now let us find in which column having the null values**

In [91]:
df.isnull().sum()

ORDERNUMBER            0
QUANTITYORDERED        0
PRICEEACH              0
ORDERLINENUMBER        0
SALES                  0
ORDERDATE              0
STATUS                 0
QTR_ID                 0
MONTH_ID               0
YEAR_ID                0
PRODUCTLINE            0
MSRP                   0
PRODUCTCODE            0
CUSTOMERNAME           0
PHONE                  0
ADDRESSLINE1           0
ADDRESSLINE2        2521
CITY                   0
STATE               1486
POSTALCODE            76
COUNTRY                0
TERRITORY           1074
CONTACTLASTNAME        0
CONTACTFIRSTNAME       0
DEALSIZE               0
dtype: int64

**In above dataset ADDRESSLINE2 consists of 2521 null values, STATE consists of 1486 null values, TERRITORY consists on 1074 null values, POSTALCODE consists of 76 null values we need to clean that null values**    

In [92]:
df["TERRITORY"]=df["TERRITORY"].fillna(df["TERRITORY"].mode()[0])
df["POSTALCODE"]=df["POSTALCODE"].fillna(df["POSTALCODE"].mode()[0])

In [93]:
df.drop("ADDRESSLINE2", axis=1,inplace=True)
df.drop("STATE", axis=1,inplace=True)

In [94]:
df

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0           10107               30      95.70                2  2871.00   
1           10121               34      81.35                5  2765.90   
2           10134               41      94.74                2  3884.34   
3           10145               45      83.26                6  3746.70   
4           10159               49     100.00               14  5205.27   
...           ...              ...        ...              ...      ...   
2818        10350               20     100.00               15  2244.40   
2819        10373               29     100.00                1  3978.51   
2820        10386               43     100.00                4  5417.57   
2821        10397               34      62.24                1  2116.16   
2822        10414               47      65.52                9  3079.44   

      ORDERDATE    STATUS  QTR_ID  MONTH_ID  YEAR_ID  ...  \
0    2003-02-24   Shipped       1         2     2003  ...   
1    2003-05-07   Shipped       2         5     2003  ...   
2    2003-07-01   Shipped       3         7     2003  ...   
3    2003-08-25   Shipped       3         8     2003  ...   
4    2003-10-10   Shipped       4        10     2003  ...   
...         ...       ...     ...       ...      ...  ...   
2818 2004-12-02   Shipped       4        12     2004  ...   
2819 2005-01-31   Shipped       1         1     2005  ...   
2820 2005-03-01  Resolved       1         3     2005  ...   
2821 2005-03-28   Shipped       1         3     2005  ...   
2822 2005-05-06   On Hold       2         5     2005  ...   

                  CUSTOMERNAME             PHONE  \
0            Land of Toys Inc.        2125557818   
1           Reims Collectables        26.47.1555   
2              Lyon Souveniers  +33 1 46 62 7555   
3            Toys4GrownUps.com        6265557265   
4     Corporate Gift Ideas Co.        6505551386   
...                        ...               ...   
2818     Euro Shopping Channel    (91) 555 94 44   
2819   Oulu Toy Supplies, Inc.        981-443655   
2820     Euro Shopping Channel    (91) 555 94 44   
2821              Alpha Cognac        61.77.6555   
2822         Gifts4AllAges.com        6175559555   

                       ADDRESSLINE1           CITY POSTALCODE  COUNTRY  \
0           897 Long Airport Avenue            NYC      10022      USA   
1                59 rue de l'Abbaye          Reims      51100   France   
2     27 rue du Colonel Pierre Avia          Paris      75508   France   
3                78934 Hillside Dr.       Pasadena      90003      USA   
4                   7734 Strong St.  San Francisco      28034      USA   
...                             ...            ...        ...      ...   
2818             C/ Moralzarzal, 86         Madrid      28034    Spain   
2819                    Torikatu 38           Oulu      90110  Finland   
2820             C/ Moralzarzal, 86         Madrid      28034    Spain   
2821          1 rue Alsace-Lorraine       Toulouse      31000   France   
2822             8616 Spinnaker Dr.         Boston      51003      USA   

     TERRITORY CONTACTLASTNAME CONTACTFIRSTNAME DEALSIZE  
0         EMEA              Yu             Kwai    Small  
1         EMEA         Henriot             Paul    Small  
2         EMEA        Da Cunha           Daniel   Medium  
3         EMEA           Young            Julie   Medium  
4         EMEA           Brown            Julie   Medium  
...        ...             ...              ...      ...  
2818      EMEA          Freyre            Diego    Small  
2819      EMEA       Koskitalo           Pirkko   Medium  
2820      EMEA          Freyre            Diego   Medium  
2821      EMEA          Roulet          Annette    Small  
2822      EMEA         Yoshido             Juri   Medium  

[2823 rows x 23 columns]

**Frome above dataset we have removed and replaced the null values ADDRESSLINE2, STATE the both the columns are remove because they havimg more than 50% of values are null values so they are not much use full for the future data visualization. TERRITORY and POSTALCODE are replaced with mode values because they are categorical data. where as categorical data is replaced with mode value and numerical data is replaced with mean and median. where as we are not using mean and median because we don't have missing values in numerical datatype** 

# **Handling Duplicate values**

In [95]:
print(df.duplicated().sum())

0


**There are no duplicate values in the dataset**